File: combination-sum/Solution39.java
// https://leetcode.com/problems/combination-sum/

import java.util.*;

class Solution39 {

    // T: O(2^t), S: O(t)
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        var result = new ArrayList<List<Integer>>();
        // Sort the array to facilitate the pruning of branches
        Arrays.sort(candidates);
        backtrack(candidates, target, 0, 0, new ArrayList<>(), result);
        return result;
    }

    // T: O(2^t), S: O(t)
    private void backtrack(int[] candidates, int target, int total, int start, List<Integer> curr,
            List<List<Integer>> result) {
        // If the current total equals the target, add the combination to the result
        if (total == target) {
            result.add(new ArrayList<>(curr));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            // If adding the current number exceeds the target, break the loop
            if (total + candidates[i] > target) {
                break;
            }
            // Add the current number to the combination
            curr.add(candidates[i]);
            // Recur with updated total and current list
            backtrack(candidates, target, total + candidates[i], i, curr, result);
            // Remove the last added number to backtrack
            curr.remove(curr.size() - 1);
        }
    }

    public static void main(String[] args) {
        var sol = new Solution39();

        var candidates1 = new int[] { 2, 3, 6, 7 };
        int target1 = 7;
        var result1 = sol.combinationSum(candidates1, target1);
        System.out.println(result1); // [[2, 2, 3], [7]]

        var candidates2 = new int[] { 2, 3, 5 };
        int target2 = 8;
        var result2 = sol.combinationSum(candidates2, target2);
        System.out.println(result2); // [[2, 2, 2, 2], [2, 3, 3], [3, 5]]

        var candidates3 = new int[] { 2 };
        int target3 = 3;
        var result3 = sol.combinationSum(candidates3, target3);
        System.out.println(result3); // []
    }

}

File: combination-sum/solution39.go
// https://leetcode.com/problems/combination-sum/

package main

import (
	"fmt"
	"sort"
)

// T: O(2^t), S: O(t)
func combinationSum(candidates []int, target int) [][]int {
	var result [][]int
	sort.Ints(candidates)
	backtrack(candidates, target, 0, 0, []int{}, &result)
	return result
}

// T: O(2^t), S: O(t)
func backtrack(candidates []int, target, total, start int, curr []int, result *[][]int) {
	if total == target {
		combination := make([]int, len(curr))
		copy(combination, curr)
		*result = append(*result, combination)
		return
	}

	for i := start; i < len(candidates); i++ {
		if total+candidates[i] > target {
			break
		}
		curr = append(curr, candidates[i])
		backtrack(candidates, target, total+candidates[i], i, curr, result)
		curr = curr[:len(curr)-1]
	}
}

func main() {
	candidates1 := []int{2, 3, 6, 7}
	target1 := 7
	result1 := combinationSum(candidates1, target1)
	fmt.Println(result1) // [[2, 2, 3], [7]]

	candidates2 := []int{2, 3, 5}
	target2 := 8
	result2 := combinationSum(candidates2, target2)
	fmt.Println(result2) // [[2, 2, 2, 2], [2, 3, 3], [3, 5]]

	candidates3 := []int{2}
	target3 := 3
	result3 := combinationSum(candidates3, target3)
	fmt.Println(result3) // []
}
